Functionalization of SuperCAT

FredericIV opened the pull request:
Rewrote SuperCAT to utilize functions for increased modularity and ease of understanding. No substantial changes have been made to collection artifacts at this time. See ReleaseNotes.txt for a complete list of changes.

BLUF:
- Regressed #42 pending reimplementation.
- Added UI framework.
- Provided for time updates to correct errant clocks.
- Changed naming and folder scheme for clarity.
- Switched from json to xml for better object handling.
- Implemented all previous pulls where applicable.

FredericIV committed:
Add files via upload

FredericIV committed:
Merge pull request #1 from FredericIV/FredericIV-Functionalization

Initial Movements towards Functions

FredericIV committed:
Revert "Initial Movements towards Functions"

FredericIV committed:
Merge pull request #2 from FredericIV/revert-1-FredericIV-Functionalization

Revert "Initial Movements towards Functions"

FredericIV committed:
Merge branch 'SuperCATDevelopers:main' into main

FredericIV committed:
Re-implemented SuperCAT to utilize functions and modules. Switched to
XML for config file. Regressed #42 Commandline arguments pending
reimplementation. See ReleaseNotes.txt for more.

FredericIV committed:
Update ReleaseNotes.txt

Update version numbering.

FredericIV committed:
Bug-Fixes

Fixed support.psm1 line ending; Updated Import-Module to align w/ current directory schema; misc bug fixes.

FredericIV committed:
Move Classification and SystemType to function

FredericIV committed:
Fixing log filenames

FredericIV committed:
Bug fixes and cleanup.

FredericIV committed:
Implement issue #52

FredericIV committed:
Merge branch 'main' of https://github.com/FredericIV/SuperCAT

FredericIV committed:
Change system time instead of delta

FredericIV committed:
Restructure files & Provide common name scheme

FredericIV committed:
Wait for SCAP and AV to finish before exiting

Handle running at midnight
General bugfixes and clarification
Begin UI tidy, aka war on new lines

FredericIV committed:
V3.3

Handle exiting, multiple selection, and reselection.

FredericIV committed:
Remove element of dev test env

FredericIV committed:
Implemented Issue #43

FredericIV committed:
Change function nouns to singular

FredericIV committed:
Prepared for merge w/ mainline SuperCat

- Changed config.xml version to fit Semantic Versioning, v0.2.2.
- Updated ReleaseNotes.txt.
- Removed development environment artifact.
- Implemented issue #43, write out drive serial number.
- Changed function nouns to singular iaw PS style guide.

lordneeko commented on Tool/Modules/Support.psm1:
```diff
@@ -0,0 +1,149 @@
+
+    if (!$NoPrompt) {
+        Write-Host
+        Set-TimeZone -Id "UTC"
```
Support.psm1 line 13 (Set-TimeZone) actually changes the timezone ON the user's machine. This needs to be changed.

FredericIV replied:
Switching the time zone before setting the time ensures that the MX guys and Operators aren't confused by the somewhat random state the time zone tends to be set in. The laptops default to mountain or central time, so setting the time will result in the clock looking substantially incorrect for most users.  -- Added a prompt notifying the user that the TZ was set.

lordneeko replied:
Adding the prompt is a good add. Still need to get buy-in from ATS that this is an acceptable action. Will look for that confirmation outside these channels.

FredericIV replied:
To allow the rest to get pulled, would it help if I split this out into a separate issue and pull request?

lordneeko replied:
No, let's leave it in and move forward. If we get information otherwise, we'll change it.

lordneeko commented on Tool/SuperCAT.ps1:
```diff
@@ -1,398 +1,359 @@
+            LastAccessTimeUTC   = Get-Date
+            LastWriteTimeUTC    = Get-Date
+            LastHDD             = ''
+            BaseName            = Read-Host -Prompt "What is the base's name"
```
We want to avoid anything that is DoD specific when we can. Keep the script prompts generic. "What is the location's name"?

FredericIV replied:
Switched prompt to read "Which campus is this?"

FredericIV replied:
Switched to "location"

lordneeko commented on Tool/SuperCAT.ps1:
```diff
@@ -1,398 +1,359 @@
+            LastWriteTimeUTC    = Get-Date
+            LastHDD             = ''
+            BaseName            = Read-Host -Prompt "What is the base's name"
+            ScanningOrg         = Read-Host -Prompt "What is the scanning organization (i.e. 52CS)"
```
We want to avoid anything that is DoD specific when we can. Keep the script prompts generic. Remove reference to 52CS

FredericIV replied:
Prompts and variables changed.

lordneeko commented on Tool/Modules/GeneralCollection.psm1:
```diff
@@ -0,0 +1,85 @@
+
+    if (!$(Test-Path $($LogPrefix | Split-Path))) { New-Item -ItemType Directory -Path $($LogPrefix | Split-Path) | Out-Null }
+    $Win32OS            = Get-WMIObject -Class Win32_OperatingSystem
+    $BaseName           = $Config.BaseName
```
Refactor all references to "Base" to simply "Location"
We want to avoid anything that is DoD specific when we can. Keep the script prompts generic.

FredericIV replied:
Refactored.

lordneeko commented on Tool/SuperCAT.ps1:
```diff
@@ -1,398 +1,359 @@
     }
-    elseif($Location -eq 4){
-        Add-Content -value "Location: $Location4" -Path "$GatherLogs\$ComputerName-Info.txt"
+    function Read-SystemVersion {
```
Need Information on what this is. MajorMinor of what? System's utilize CPINs for versioning of software.

FredericIV replied:
CPINs, being between 17 and 39 characters long are impractical to stuff into a file name.  Do we feel that having quick reference as to the version of any part of the system is valuable when performing system analysis? I'm not tied to the MajorMinor format, or even the existence of the field.

lordneeko replied:
I agree CPIN is too long, however, I'm not sure what most users would put here.  Do the systems we're looking at prototyping this on have a Major/Minor label?

FredericIV replied:
I known the system we have the dev build for does, but I have no knowledge of the other systems. I would expect them to though, as it's the standard model for software versioning (Major.Minor.Bugfix rRelease)

lordneeko replied:
We can leave it for now... I'd rather address a file naming solution.  In addition, in the future, I think some of these user entered details should make their way into the -info.csv file instead.

lordneeko commented on Tool/PlatformList.csv:
```diff
@@ -0,0 +1,6 @@
+SystemAbbreviation,SystemName
```
Remove these system names and put generic "System A," System B" for the github repo.

FredericIV replied:
Removed systems and added generic system names.

lordneeko commented on Tool/SuperCAT.ps1:
```diff
@@ -1,398 +1,359 @@
-    $Choices = Read-Host "Please only input numbers 0 to 7 and commas (i.e. 1,3,5)"
-}
+######################### Module Imports #######################################
+using namespace System.Collections.Generic
```
**Message from PSScryptAnalyzer**
[{
	"resource": "/FredericIV/SuperCAT/Tool/SuperCAT.ps1",
	"owner": "_generated_diagnostic_collection_name_#0",
	"code": "PSUseCompatibleSyntax",
	"severity": 8,
	"message": "The using statement syntax 'using ...;' is not available by default in PowerShell versions 3,4",
	"source": "PSScriptAnalyzer",
	"startLineNumber": 10,
	"startColumn": 1,
	"endLineNumber": 10,
	"endColumn": 43
}]

FredericIV replied:
Removed System.Collections.Generic dependency. Had to wrangle some weird typecasting bugs, but it's done.

lordneeko reviewed (changes requested):
Most significantly, please "genericize" the script and remove references to DOD and DAF. Other items up for discussion.

FredericIV committed:
Cleanse of references

Remove references and clean up prompts

FredericIV committed:
Bugfixes

Typecasting is awful.
Fixed a time related bug where it wouldn't translate the correct time to the current time zone.
Auto-pad drive number to three digits.
Added a few function synopsis.
Removed System.Collections.Generic dependency.
Fixed drive number and version number validation.

FredericIV committed:
UTC and Location

UTC: Time to die on a hill.

FredericIV committed:
Squashed commit of the following:

commit 08efea4735a1f2c0c8f8356c79299dfea387f756
Author: Frederic Borries IV <frederic@fabiv.pw>
Date:   Tue Jan 11 12:43:05 2022 +0100

    UTC and Location

    UTC: Time to die on a hill.

commit 897f3a9f98d2c52921b65cfd29fac262735b1a3d
Author: Frederic Borries IV <frederic@fabiv.pw>
Date:   Tue Jan 11 00:21:41 2022 +0100

    Bugfixes

    Typecasting is awful.
    Fixed a time related bug where it wouldn't translate the correct time to the current time zone.
    Auto-pad drive number to three digits.
    Added a few function synopsis.
    Removed System.Collections.Generic dependency.
    Fixed drive number and version number validation.

commit b57e69a6c80d712df8e09051de26d9cf5f6241b7
Author: Frederic Borries IV <frederic@fabiv.pw>
Date:   Mon Jan 10 22:22:33 2022 +0100

    Cleanse of references

    Remove references and clean up prompts

FredericIV committed:
Merge branch 'main' into main

lordneeko commented on Tool/SuperCAT.ps1:
```diff
@@ -1,400 +1,345 @@
+    if ($SCAP) {
+        return "$SystemOwner`_$SystemAbbreviation$Version`_$Drive"
+    } else {
+        return "$Date`_$SystemOwner`_$SystemAbbreviation$Version`_$Drive"
```
We really need to discuss this naming convention. I get the reasoning, and I do not yet have a great recommendation that would work for all systems...but as someone who handles data analysis...I try my best to stay away from user entered data.  Data that is collected automatically and logged is more repeatable.
Also, I've looked through the code (and my own runs of the code) and I do not know what "systemowner" is. Also, the files that it output didn't appear to have that on it.

What is the reason we cannot use 
$datetime_$SystemName_$HardDriveModel_$HardDriveSerial
In this instance, the only thing user entered is the $SystemName and the HDSerial should ensure it is unique.

FredericIV replied:
Agreed that we need to nail down the file name scheme. SystemOwner in this case is who owns/uses this HDD. Separate from ScanningOrg, which is the organization in charge of monitoring and analysis. SystemName(/Type), SystemAbbreviation, and Version should be pull-able from the system, just not in a format available on GitHub. This leaves SystemOwner and DriveName to user entry. The reason we used DriveName, which is actually a padded three digit number, is for brevity and consistency in the file name scheme. System owner could theoretically be a drop down, but that would require constant monitoring and updating.  The reason for not using $HardDriveModel_$HardDriveSerial is enabling instant comprehension by both security personnel and the system owners.

As an aside, SCAP has its own folder structure we're not messing with to allow for usage of existing analysis programs.

lordneeko replied:
"The reason for not using $HardDriveModel_$HardDriveSerial is enabling instant comprehension by both security personnel and the system owners."

I would think that would drive us to use Model_SerialNumber instead, since those drives will (should) have a sticker on the outside of them that would correspond to the filename.

Is "SystemOwner" a new prompt that was added to the setup script part of the script? I haven't run your latest in a few days, but don't remember being prompted for that when I ran it.

FredericIV replied:
SystemOwner is the genericized Unit, but the meaning of the question is the same.  The thought behind DriveName is that the system owners have these numbered already for tracking purposes (i.e. Laptop 1 w/ Drive 5 checked out).

lordneeko reviewed (changes requested):
Let's just do some testing and finalize the file naming. I really want to get this right now so we don't have to change it in the future (because that REALLY messes up our automated data analysis management)

FredericIV committed:
Merge pull request #3 from FredericIV/Corrections

Corrections

FredericIV committed:
Fix abbreviations in event of "Other"; Add classification abbreviation
